README.md
# blazars-mcmc

blazars-mcmc contains code to use MCMC to model the spectral energy distributions of blazars. It can be used for a synchrotron-self-Compton model and for a model that is SSC + external inverse-Compton.

## To Use:
By default, results are written to a file in `local_results` entitled `run_yyyy-mm-dd-hh:mm:ss` and configurations are read from `mcmc_config.txt`. 
1. Ensure all dependencies are installed, see [dependencies](#dependencies). Recommended: create conda env from `environment.yml` using `conda env create -f environment.yml`
2. Create a copy of `mcmc_config_template.txt` called `mcmc_config.txt`. For information, see [configuration file](#mcmc-configuration-file). 
3. Ensure the data file is formatted as described below and is at the relative path specified in `mcmc_config.txt`, see [data format](#data-format).
4. Navigate inside folder `blazar_mcmc`. Currently, code **must** be run from inside this folder or the imports won't work.
5. First time use: execute `blazar_initialize.py` with `python blazar_initialize.py` or `python3 blazar_initialize.py` depending on your Python setup. This creates all necessary folders and compiles the C++ code (see [bjet](#using-bjet_02)).
6. Execute `blazar_run_mcmc.py` with `python blazar_run_mcmc.py` or `python3 blazar_run_mcmc.py` depending on your Python setup. 

`console_data_processing.py` can be executed in an interactive Python console. It loads all information necessary for calling functions to create statistics and plots given a path relative to `BASE_PATH` for a results folder. 

## Setup
### Dependencies
###### **Recommended:** create conda env from `environment.yml` using `conda env create -f environment.yml`


`emcee`
- To install with conda: `conda install -c conda-forge emcee`
- To install with pip: `emcee` recommends calling `pip install -U setuptools setuptools_scm pep517` and then `pip install -U emcee`
- `emcee` installation [documentation](https://emcee.readthedocs.io/en/stable/user/install/).

`scipy`
- To install with pip: `python -m pip install scipy`
- To install with conda: `conda install -c conda-forge scipy`
- To install with pip: `python -m pip install -U scipy`
- `scipy` installation [documentation](https://scipy.org/install/)

`invoke`
- To install with pip: `python -m pip install invoke`
- To install with conda: `conda install -c conda-forge invoke`


`matplotlib`
- To install with conda: `conda install matplotlib`
- To install with pip: `python -m pip install -U matplotlib`
- `matplotlib` installation [documentation](https://matplotlib.org/stable/users/installing/index.html)

`corner`
- To install with conda:`conda install -c astropy corner`
- To install with pip: `python -m pip install corner`
- `corner` installation [documentation](https://corner.readthedocs.io/en/latest/install.html)

`tqdm`
- To install with pip: `python -m pip install tqdm`

`h5py`
- `python -m pip install h5py`

**NOTE:** `numpy` is also necessary. The `emcee` installation will install `numpy` if it is not there already. 
`numpy` itself can be installed:
- Using conda: `conda install numpy`
- Using pip: `pip install numpy`

### Necessary Contents:
- Directory `Bjet_02` (do not modify except to edit Makefile if necessary)
- Directory `sed_calculations` (do not modify or delete)
- Directory `mcmc_results`, where data will go (do not delete directory)
- Directory `parameter_files` (do not modify or delete)
- `blazar_run_mcmc.py`, `blazar_model.py`, `blazar_plots.py`, `blazar_utils.py`, `blazar_properties.py`, `blazar_report.py`

## Formatting information
### Listing parameters
Parameters are always listed in the following order:
[delta, K, n1, n2, gamma_min, gamma_max, gamma_break, B, R]
The additional parameters for EIC are bb_temp, l_nuc, tau, blob_dist, in that order.
All parameters are the logarithm of the true value except for delta, n1, and n2

---------------------------------------------------
| Variable    | Description                 | Scale  |
|-------------|-----------------------------|--------|
| delta       | Doppler factor              | linear |
| K           | particle density [cm^-3]    | log    |
| n1          | alpha_1 (first index)       | linear | 
| n2          | alpha_2 (second index)      | linear | 
| gamma_min   | low-energy cutoff           | log    | 
| gamma_max   | high-energy cutoff          | log    | 
| gamma_break | energy break                | log    | 
| B           | magnetic field strength [G] | log    | 
| R           | blob radius [cm]            | log    | 
---------------------------------------------------
| Additional for EIC | Description                  | Scale |
|--------------------|------------------------------|-------|
| bb_temp            | Black body temp of disk [K]  | log   | 
| l_nuc              | Nucleus luminosity [ergs/s]  | log   | 
| tau                | Frac of luminosity scattered | log   | 
| blob_dist          | Distance of blob [cm]        | log   | 

### Data format
Data by default is assumed to be in data files with the following columns as labels in the first row:
`!E(eV)		F(ergcm-2s-1)	delta e(-)	delta e(+)	delta F(-)	delta F(+)	Observatory`.
The first row is therefore skipped when reading data.

`E`, `F`, and `delta F(-)` are the columns used by the program, which by default are in columns 0, 1, and 4.
As an optional parameter, these can be changed. The data must be space-delimited.

All rows of data **must have the same number of rows**.

### MCMC configuration file
The default configuration file is `mcmc_config.txt` in the main directory. If you would like to use a different file, in `blazar_run_mcmc.py` in main, replace `config_file=None` and/or `directory=None` with the location of the config file and/or the desired directory for results. (If the defaults are OK, no code should need editing.)

There is a file named `mcmc_config_template.txt` as an example. Make a copy of this file named `mcmc_config.txt`. 
This file is automatically in `.gitignore` since it changes locally.

Rows are:
```
description = <description of run>
folder_label = <prefix for folder name>
eic = <True/False, SSC or SSC + EIC>
data_file=<file name relative to blazars-mcmc>
n_steps=<# of steps>
n_walkers=<# of walkers>
discard=<steps to discard>
parallel=<True or False (use parallel processing)
cores = <# of cores to use if parallel True, optional>
tau_var=<Tau variability in hours>
redshift = <float for redshift>
custom_alpha2_limits=False or <True, lower, upper>

```
Notes: 
- Labels must be exactly as listed
- `data_file` is the *relative path* from the home directory to the file with data.
#### Example mcmc configuration file
(file named `mcmc_config.txt`, can be modified)
```
# Configuration file for running mcmc

# description:
description=3C66A b2 eic

# folder label:
folder_label =3C66A_b2_eic

# eic (True/False):
eic=True

# Data file:
data_file=real_data/3C66A_sed_block2.dat

# Number of steps:
n_steps=4000

# Number of walkers:
n_walkers=300

# Discard number:
discard=500

# Parallel processing (True/False):
parallel=True

# If parallel processing, # of cores (will use # of cores - 1 if not specified)
#cores=9

# use tau variability (boolean):
use_variability= True

# tau variability (in hours):
tau_variability = 168

# redshift (J1010 0.143, 3C66A 0.340)
redshift = 0.340

# Custom alpha2 limits (True/False, <val1>, <val2>) val1 and val2 optional
custom_alpha2_limits=False
```
## Code information
The `blazar_run_mcmc.py` script creates a results folder, by default in local_results named with the folder_prefix and the current date.

At the start, a `basic_info.txt` doc is created in the folder with configuration information saved.

The history of the entire chain from the MCMC and the corresponding log_probs are stored in an h5 file named `backend.h5`. This is updated continuously. This means that if the code crashes, all existing results are still present. h5 files can be re-loaded as the current state for the MCMC at the beginning of another run (see the main method in `blazar_run_mcmc.py`) or loaded to use for statistics or plotting (see `load_backend()` in `blazar_report.py`).
### Using the blazar_mcmc code
Using the functions directly rather than running the `blazar_run_mcmc.py` file as a script allows for greater customizability and increased functionality.

See docstrings of the modules for function information and the docstrings of individual functions for more detailed information. 


General note: paths are generally specified relative to the base directory (`blazars-mcmc`) or in more specific function to the parent folder (`RESULTS_FOLDER` or `DATA_FOLDER`). See function documentation.
#### Modules:
- `blazar_clean.py`: Deletes old model data files
- `blazar_initialize.py`: Compiles the c++ code, ensures all necessary directories are present. 
- `blazar_model.py`: Functions for creating models
- `blazar_plots.py`: Functions for plotting
- `blazar_properties.py`: Advanced configuration options and default values
- `blazar_report.py`: Functions for calculating stats and creating reports
- `blazar_run_mcmc.py`: Main python file for running the MCMC.
- `blazar_utils.py`: Utility functions

#### Notable Options
In the main method in `blazar_run_mcmc.py`, it is possible to specify the starting position (p0) for the MCMC. This is useful for continuing a run after a crash.

The `save()` function in `blazar_report.py` will create all plots and statistics given just the name of the directory that contains an h5 file and an `info.txt` doc. 

*Make models, plot results, and get statistics either in the Python console or in scripts:* 
- `blazar_report.parse_info_doc()` will, given a path to an `info.txt` file or `basic_info.txt` file, parse the file and return a dictionary with all information. Notably, `info["configs"]` contains the dictionary of configurations.
- `blazar_utils.read_configs()` will, given the path to an `mcmc_config.txt` (default is `"mcmc_config.txt"`), return a dictionary of configurations.
- `blazar_utils.read_data()` returns a tuple of numpy arrays with the `v_data`, `vFv_data`, and the `err_data` given a data file.
- `blazar_report.load_from_backend()` will return a tuple of the entire chain from the MCMC and the log_probs given a data folder relative to `RESULTS_FOLDER`
- After using the above in a script or from the console, statistics and plots can be calculated from functions in `blazar_plots.py` and `blazar_report.py`. 
## Using bjet_mcmc

`bjet_mcmc` contains a modified version of the Bjet_02 code written by Olivier Hervet that is used for the MCMC. 

Compile the exec with `make bj_mcmc`.

Code is in the `bjet_mcmc` folder. The necessary files are `bj_mcmc.cpp`, `bj_mcmc.h`, `processes_supp_mcmc.cpp`, and `processes_supp_mcmc.h`
### Usage
#### Options for calling the executable:
- Usage 1: `bj_mcmc --help`
- Usage 2: `bj_mcmc <parameter file>`
- Usage 3: `bj_mcmc 0 <parameter file>` same execution as usage 2, `_prev files` made
- Usage 4: `bj_mcmc 1 <parameter file>`                           no _prev files made
- Usage 5: `bj_mcmc 2 <data folder> <parameter file>`            _prev files made
- Usage 6: `bj_mcmc 3 <data folder> <parameter file>`               no _prev files made
- Usage 7: `bj_mcmc 0/1 <model type> <model params, at least 19 depending on model>`
    ^params from command line, 0 = yes `_prev files`, 1 = no
- Usage 8: `bj_mcmc 2/3 <data folder> <model type> <model params, at least 19 depending on model>`
    ^params from command line, 2 = yes `_prev files`, 3 = no

#### Parameter files
Parameter files specified in the command line should be paths relative to the `bj_mcmc` executable. See `params.txt` in `bjet_mcmc` for an example parameter file.
#### Modes:
- 0: `_prev` files are made, save in default data folder
- 1: no `_prev` files are made, save in default data folder
- 2: `_prev` files are made, save in specified data folder
- 3: no _prev files are made, save in specified data folder

(default data directory is in the same directory as the executable, named "data")

#### Valid values for model type:
- 0: model with just blob
- 1: blob + EIC parameters
- Other models not yet implemented

#### Order of parameters in command line:
##### Model type 0:
 ```
 bj_mcmc <0, 1, 2, 3> <data folder if applicable> 0 (model type)
 z, H_0, THETA, DOP_B, K1, N1, N2, GAMMA_MIN, GAMMA_MAX, GAMMA_BRK, B, R_src,
 L_src, IIR_level, D_b, NU_DIM, NU_STR, NU_END, prefix
 ```
 argc should be 22 or 23 depending on if data folder is listed

##### Model type 1:
 ```
 bj_mcmc <0, 1, 2, 3> <data folder if applicable> 1 (model type)
 z, H_0, THETA, DOP_B, K1, N1, N2, GAMMA_MIN, GAMMA_MAX, GAMMA_BRK, B, R_src,
 L_src, IIR_level, D_b, T_BB, TBB_tor, L_nuc, tau, L_tor, tau, NU_DIM, NU_STR, NU_END, prefix
 *Note that tau is present twice, this is a slight error in the bjet code. The second tau value is not used for
 anything, but it must be inputted.
```
 argc should be 28 or 29 depending on if data folder is listed

Example:
```
 ./bj_mcmc 3 /Users/sed_calculations 1 0.34 69.6 0.57 50.0 612.1 2.28 3.74 2816.9 1803000 44806 0.00236 5.94e+17 0 1 3.8e+15 2013 2.0e+4 1.7e+21 1.5e-10 5.5e+20 9.0e-5 99 50000000.0 1e+29 run
```
 ^ here, the 3 indicates that the data folder is specified and no prev file is made. 1 is the EIC model type. Then 0.34 is z (redshift) and then the rest of the parameters are enumerated.


log/CHANGELOG.md
#CHANGELOG
###last updated: 2022-05-30

**2022-05-30** (27ca2c1) HUGE ERROR found. When make_model called
process_model, it did not specify EIC, making the EIC component do
practically nothing. This was really bad.     

     Plotting functions in plot_data_with sources modified to make
     results for the paper.

**2022-05-28** (604e45e) Fixed big in plotting chi squared, added
options for parsing past info.txt files and processing backends.     

     Previously, the plotting the chi squared directly modified the
     array instead of making a copy, which meant that the wrong plots
     were created. info.txt and basic_info.txt files can be parsed to
     read the configurations. This combined with h5 files means that
     data can be re-loaded with the configurations used when the code
     was run. There is now a save function in blazar_report that allows
     just inputting a folder instead of having to directly pass all the
     function results like indices within one sigma. Code was written in
     plot_data_with_sources.py to plot the different sources of the
     3C66A block 1 data in different colors. (This will probably work
     for other blocks too, but I just eyeballed the number cutoffs for
     values that worked with the data, so this may not be true.)

**2022-05-25** (31a1fe6) Delete Bjet_02_2022_03_28.tar.xz     


**2022-05-25** (fcad12e) Add files via upload     


**2022-05-24** (d9e03b7) Changelog updated     


**2022-05-24** (b8ac193) mkdtemp used instead of TemporaryDirectory to
prevent being deleted prematurely with nohup. blazar_run.sh runs
caffeinate only on Darwin and not Linux.     


**2022-05-19** (81ba653) blazar_run.sh runs the command to run the
program and saves the process number. Output of compiling bjet is now
suppressed by default in blazar_initialize     


**2022-05-19** (6c3ceaa) 3C66A blocks 2, 5, 6 data added     


**2022-05-19** (9e37a3e) Removed excess error messages from Makefile     


**2022-05-19** (943c2b1) TMP setting now in blazar_properties. This
makes files saved in temp directory.     


**2022-05-13** (9c4123c) Make compression at end of mcmc cross-platform     


**2022-05-13** (575ba4f) Changed USE_BASE_PATH back to false to prevent
problems     


**2022-05-13** (56295ff) Add sed_calculations to gitignore     


**2022-05-13** (3c545cd) Descriptive names to results files, remove h5
and pickle files now stored on google drive to decrease size     


**2022-05-13** (f50913e) Remove circular dependency in blazar_model and
blazar_utils, change Hubble constant     

     blazar_model now contains the log_to_linear and the linear_to_log
     functions so that it does not need to reference blazar_utils.

     Hubble constant is now 69.6, not 71.0

**2022-05-13** (b7c739c) EIC results for 3C66A added from VHE     


**2022-05-07** (c3db53b) mcmc_config is now in .gitignore,
mcmc_config_template.txt is added as a guide     


**2022-05-07** (931f663) Fixed read_configs to allow for no
tau_variability provided     


**2022-05-07** (d3a3f4f) Add eic to configs file     


**2022-05-07** (ccd76d0) Merge branch 'main' of
github.com:sarah829/blazars-mcmc into main     

     branch from vhe

**2022-05-07** (f74bc2e) requirements.txt from pip environment created     


**2022-05-07** (37ae0b8) Fix error in plot_1sigma, add redshift and
use_variability to configs, add redshift as optional param to relevant
functions, optional parameters added     

     In get_params_1sigma_ranges, minima and maxima started
     as all copies of flat_samples[0], which was a problem because that
     is not necessarily within 1 sigma. It now starts with
     flat_samples[indices_within_1sigma[0]], which fixes the error. 
     Redshift is now in the parameter file and is passed into the
     log_probability function, so it was added to the chi_squared
     function, etc. Having a variability constraint is now optional and
     set by use_variability in the config file. Functions now have more
     of the optional parameters that the functions they call have,
     allowing for more flexibility. Code was re-formatted.

**2022-05-05** (fe4d13d) Typo fixed (added space before Todo)     


**2022-05-05** (017a3cd) Changed name stem used in make_model in
blazar_report, fixed typo in README     

     make_model in blazar_report now uses run<random number> for the
     file name stem, and the files are deleted after use. Additionally,
     it no longer uses a parameter file (now the MCMC code never uses
     one).

**2022-05-05** (a12213d) Add todo list to README, update name stems in
blazar_plots, add 3C66A data     

     Name stems in blazar_plots are now for_plotting<random number>
     instead of just for_plotting. This allows for running multiple runs
     of the program at the same time without problems. The functions now
     delete the created SED models after running.

**2022-04-27** (ce0b76f) readme updated to reflect changes in file
organization     


**2022-04-23** (ade3ddc) References to blazar_eic.py removed     


**2022-04-22** (81dc301) Create local_results and old directories and
add to .gitignore     

     local_results contains results files that don't need to be in the
     public repo, and old contains now defunct code and results. Files
     in .gitignore were dropped.

**2022-04-22** (7f65dd8) File reorganization starting to clear out old
files (results, defunct test scripts, etc.) from the repo     


**2022-04-22** (2ace9f0) Merged eic branch to main     


**2022-04-21** (bf80b56) Removed folder for gitignored backups     


**2022-04-21** (b48dafa) EIC code now integrated with regular code
instead of in blazar_eic.     

     Whether or not the model is an eic model is stored in a boolean eic
     for most methods. Additionally, in main of blazar_run_mcmc, the h5
     and pickle files are zipped.

**2022-04-03** (1664363) MCMC with EIC now works, BlazarProperties class
added.     


bjet_mcmc/bj_mcmc.h
/*
Slightly modified version of bj02.h for use in MCMC code; modified by Sarah Youngquist in Feb 2022
See bj_mcmc.cpp for an explanation of the changes.
*/

#ifndef BLAZARS_MCMC_BJ02_COPY_H
#define BLAZARS_MCMC_BJ02_COPY_H


using namespace std;

#include <time.h>

namespace bj_mcmc02 {



// FLAGS
    const int PRINT = 0; // 1=True, 0=False
    const int DEBUG = 0; // 1=True, 0=False
//#warning "DEUBG flag enabled"

// PHYSICAL CONSTANTS c.g.i.

    const double q0     = 0.5;
    const double pc     = 3.086    * 1.0e+18; // [cm]
    const double H0SI   = 70.;                // SI:  km/s/Mpc
    const double H0     = H0SI*1.e5/(1.e6*pc);// cgs: s-1
    const double sig_T  = 6.652453 * 1.0e-25; // [cm^2]
    const double m_e    = 9.109558 * 1.0e-28; // [g]
    const double c      = 2.997924 * 1.0e+10; // [cm / s]
    const double h      = 6.626296 * 1.0e-27; // [erg * s]
    const double sig    = 5.66961  * 1.0e-05; // [erg / (s * cm^2 * K^4)]
    const double m_p    = 1.672623 * 1.0e-24; // [g]
    const double e      = 4.803250 * 1.0e-10; // [esu]
    const double keV    = 2.417965 * 1.0e+17; // [Hz]
    const double G      = 6.6726   * 1.0e-08; // [dyn / (g^2 * cm^2)]
    const double k_B    = 1.380622 * 1.0e-16; // [erg / K]
    const double M_sol  = 1.989    * 1.0e+33; // [g]
    const double eV_K   = 11.6048  * 1.0e+3;  // [K]
    const double hour   = 3600.0;             // [s]
    const double day    = 24.0 * hour;        // [s]


// PHYSICAL CONSTANTS S.I.

    const double PLANCK = 6.6260693e-34; // J.s
    const double eV     = 1.602e-19;     // J
    const double PARSEC     = pc;	// cm
    const double ERG        = 1.e-7;	// J
    const double MELEC      = 0.511e6;   // eV
    const double HZ_PER_EV  = 2.145e14;
    const double LIGHTSPEED = c; // cm/s
    const double SIGMAT     = sig_T; //cm2
    const double HMASS      = 1.673e-27;	// kg
    const double MSOL = 1.989e30;	// kg
    const double ALPHA = 0.0073;	// fine structure const
    const double MILLIBARN = 1.e-27; // cm^2
    const double CHARGE = 1.602e-19; //Coulomb
    const double MELECKG = m_e; //g
    const double MPION = 135.e6;  //eV
    const double MPROTON = 938.272e6;  //eV

// CALCULATION PRECISION

    const int    SYN_PREC1       = 15;      // INTEGRATION PRECISION FOR SYNCH. SPEC.
    const int    SYN_PREC2       = 15;      // INTEGRATION PRECISION FOR SYNCH. SPEC.
    const int    ABS_PREC1       = 15;      // INTEGRATION PRECISION FOR ABSOR. SPEC.
    const int    ABS_PREC2       = 15;      // INTEGRATION PRECISION FOR ABSOR. SPEC.
    const int    COM_PREC1       = 15;      // INTEGRATION PRECISION FOR I.COM. SPEC.
    const int    COM_PREC2       = 15;      // INTEGRATION PRECISION FOR I.COM. SPEC.
    const int    G_MAX           = 1000;    // MAXIMUM NUMBER OF GAMMA POINTS
    const int    NU_MAX          = 5000;     // MAXIMUM NUMBER OF SPECTRAL POINTS


// TIME VARIABLES

    extern time_t       T_START, T_END;
    extern char         PARA_FN[256];

// GLOBAL VARIABLES


    const int    G_DIM=100;     // CURRENT NUMBER OF GAMMA POINTS
    const double N_MIN=1.0e-15; // MINIMAL ELECTRONS DENSITY
    const double N_MAX=50;  // MAXIMAL ELECTRONS DENSITY

//extern char  prefix[256];

    const int    NU_DIM_MAX      = 300;
    const int    SL_DIM_MAX      = 200;


// PHYSICAL PARAMETERS

    extern double       z;
    extern double       H_0;
    extern double       THETA, D_L;
    extern double       DOP_B, LOR_B, V_B, V_B_APP;//DOP_BB,
    extern double       R_src,R_blr;
    extern double       L_src, L_nuc;
    extern double       B;
    extern double       K1;
    extern double       K2;
    extern double       N1;
    extern double       N2;
    extern double       GAMMA_MIN;
    extern double       GAMMA_BRK;
    extern double       GAMMA_MAX;
    extern double       T_BB;
    extern double       tau;  // fraction of L_nuc scattered/rerocessed isotropically (EIC)
    extern double       B_0;
    const double        n_B = 1.0; // slope of the magnetic field along the jet
    extern double       N_0;
    extern double       n_n;
    extern double       n_N;
    extern double       n_G;
    extern double       D_b, D_b_src_j;
    extern double       U_B,U_e;
    extern double       jj1,kk1;
    extern double       L_jet_eic;
    extern double       L_tor, T_BB_tor;
    extern double       Tcool_vhe, Tcool_radio;

// TRANSFORMATION PARAMETERS

    extern double DOP_JET, LOR_JET, V_JET, V_JET_APP, DOP_B_J, LOR_B_J, V_B_J;

// GEOMETRY PARAMETERS

//const int           SL_DIM_MAX = 200;
    extern int          SL_DIM;
    extern double       PHI, PHI_src;
    extern double       THETA_src_j;
    extern double       J_LEN, J_LEN_src;
    extern double       A;
    extern double       X_MIN;
    extern double       X_MAX;
    extern double       X_OUT;
    extern double       Y_MIN;
    extern double       Y_MAX;
    extern double       R_MOY;
    extern double       Xcut;
    extern double       antisym;
    extern double       Sprev;
    extern double       Sum_S;

// OTHER PARAMETERS

    extern int          SL_CUR;
    extern double       GAMMA_MAX_0;
    extern int          null0;
    extern double       V_exp, DEL_Tph, DEL_Tj;



// SUBROUTINES ******************************************************************
    extern double N_e(double gamma);


// METHODS
    void description();
    // Sarah Youngquist added
    int run_models(); // moved prev code from main here
    int load_params(char* name);
    int load_params_from_list(char** list, int model_type, int starting_index, int list_length);
};

#endif //BLAZARS_MCMC_BJ02_COPY_H
